Support rpc_async call with timeout in JIT

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37884 Support rpc_async call with timeout in JIT**

Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in `jit/rpc_test_faulty`.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

rohan-varma committed:
Support rpc_async call with timeout in JIT

Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

xush6528 commented on torch/csrc/jit/runtime/register_distributed_ops.cpp:
```diff
@@ -121,13 +122,19 @@ RegisterOperators reg_rpc_ops(
              auto& kwargsDictIValue =
                  num_inputs >= 4 ? *stackIter++ : emptyDict;
+
+            // IValue corresponding to placeholder for RPC timeout. Used if no
```
Missed a leading space?

xush6528 commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -2785,7 +2789,13 @@ struct to_ir {
     // respectively, users are allowed to not pass `args` and `kwargs`.
-    TreeList args_kwargs_trees(input_trees.begin() + 2, input_trees.end());
+
+    TreeList args_kwargs_trees(
+        input_trees.begin() + 2,
+        apply.inputs().size() == rpcMaxInputs ? input_trees.end() - 1 : input_trees.end());
+
+    // Parse timeout, if applicable
+    Value* timeout = apply.inputs().size() == rpcMaxInputs ? emitExpr(Expr(input_trees[4])) : nullptr;
```
I think you don't need to change this part.

You just need to change `args_kwargs_trees` to `args_kwargs_timeout_trees`.

rohan-varma replied:
Good point!

xush6528 commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -960,3 +990,76 @@ def test_call_fork_in_jit_with_profiling(self):
         self.assertEqual(function_event.name, "foo")
+
+
+class JitFaultyAgentRpcTest(FaultyRpcAgentTestFixture):
```
Need to add this TestCase to class JitRpcTest as well, otherwise the test is not run.

Also, no need to inherit from FaultyRpcAgentTestFixture. JitRpcTest does that for us.

rohan-varma commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -2751,9 +2751,13 @@ struct to_ir {
     // The RPC C++ entry API can take c10::Function directly.
-    if (apply.inputs().size() < 2 || apply.inputs().size() > 4) {
+    auto rpcMinInputs = 2;
+    auto rpcMaxInputs = 5;
```
Hmm not sure how to fix this clang-tidy error, or why it's erroring for 5 and not 2...

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

xush6528 reviewed (approved):
Glad to see unified API!

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in jit/rpc_test.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in `jit/rpc_test_faulty`.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support rpc_async call with timeout in JIT"


Adds support to use rpc_timeout param in rpc_async call from jit for
parity with eager mode. Done by:
1) Add timeout as an input in ir_emitter.cpp if it is specified
2) Parse float IValue from inputs in `prim::rpc_async` operator. Give the default if needed.

Added UTs in `jit/rpc_test_faulty`.

Differential Revision: [D21268895](https://our.internmc.facebook.com/intern/diff/D21268895/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21268895/)!

[ghstack-poisoned]
